New feature "Show student progress in My courses page"

hi-rafa opened the pull request:
This feature will display to the student's progress, rating and certificate below the title of each of his courses. By default the feature will be deactivated and it will be up to the admin to activate it or not.


1 - Admin: activate new option (Show course progress to student)
main/admin/settings.php?category=Course

![screen shot 2017-02-13 at 9 55 32 am](https://cloud.githubusercontent.com/assets/3813277/22876950/5ed5bca2-f1d3-11e6-9e4e-a2e4d1285509.png)


2 - Student: Progress displayed for each course
user_portal.php

<img width="1278" alt="screen shot 2017-02-13 at 10 00 53 am" src="https://cloud.githubusercontent.com/assets/3813277/22876954/6460b5a0-f1d3-11e6-9fde-f93093ee690c.png">

ywarnier commented:
Hello @hi-rafa, this is a great addition.
However, we do not accept database changes (even addition of settings) between minor versions of Chamilo, unless these are necessary to fix a migration but.
However, we do use a "trick" to bypass that until the next major version: we add the setting in the following form in main/install/configuration.dist.php:
`$_configuration['my_course_progress']` = false;`
This ensures that any new install will have the config param available in configuration.php, and that any upgraded install can just copy that line from configuration.dist.php to configuration.php and change it to "true" to enable the feature.

Inside Chamilo, you then use api_get_configuration_value('my_course_progress') to get that value (instead of api_get_setting()).
Example here: https://github.com/chamilo/chamilo-lms/commit/0cadaa026fcaf10fcb80c9c6fed0548495c04a8f#diff-9ce12ac2fc0681bc9b2bbbc5e3ddac94R1461
The api_get_configuration_value() function returns false if the variable is not defined, so it is important that the default value (to leave the behaviour of Chamilo as it was before) be false.

In this case, your choice of variable is OK, but it wouldn't be OK, for example, to have a setting called "hide_my_course_progress", because then the default value being false would mean that we *do* show the progress, which changes the default behaviour of Chamilo, etc.

hi-rafa commented:
Hi @ywarnier, 
Thank you for the trick. I did apply the changes as you mentioned above, I hope everything is ok now.

ywarnier commented:
Looks OK at code level. I will test it as soon as possible before merging.

hi-rafa commented:
In my last commit I fixed errors I made and I added a new condition. The (Progress, Rating and Certificate) will only display if they have been assigned.

<img width="1204" alt="screen shot 2017-03-22 at 10 02 53 am" src="https://cloud.githubusercontent.com/assets/3813277/24190247/611f1382-0ee8-11e7-94e7-7cd7ca5ee0a4.png">

ywarnier commented on main/inc/lib/course.lib.php:
```diff
@@ -3554,6 +3554,8 @@ public static function returnSpecialCourses($user_id, $load_dirs = false)
                 }
 
                 $params = [];
+                //Param (course_code) needed to get the student process
```
the student "progress", I guess?

hi-rafa replied:
yes sorry

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1165,11 +1165,118 @@ public function returnCoursesAndSessions($user_id)
 
+
+            //Course option (show student progress)
+            //This code will add a new variables (Progress, Score, Certificate)
```
"will add new variables" (without "a")

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1165,11 +1165,118 @@ public function returnCoursesAndSessions($user_id)
+
+            //Course option (show student progress)
+            //This code will add a new variables (Progress, Score, Certificate)
+            if(api_get_configuration_value('my_course_progress') === true && api_is_student() === true) {
```
Missing a space after "if" (please check PSR-2: "Control structure keywords MUST have one space after them" here: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md)

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1165,11 +1165,118 @@ public function returnCoursesAndSessions($user_id)
+            //This code will add a new variables (Progress, Score, Certificate)
+            if(api_get_configuration_value('my_course_progress') === true && api_is_student() === true) {
+                
+                foreach($specialCourses as $key => $specialCourseInfo) {
```
Same observation

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1165,11 +1165,118 @@ public function returnCoursesAndSessions($user_id)
+                if(isset($courses['in_category']) && isset($courses['not_category'])){
+                    foreach($courses['in_category'] as $key1 => $value) {
+                        if(isset($courses['in_category'][$key1]['courses'])){
+                            foreach($courses['in_category'][$key1]['courses'] as $key2 => $courseInCatInfo) {
```
Same spacing observation (plus same observation for the end of the line: missing space before "{") for these 4 lines

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1165,11 +1165,118 @@ public function returnCoursesAndSessions($user_id)
+                                );
+                                $courses['in_category'][ $key1 ]['student_info']['certificate'] = null;
+                                if (isset($category[0])) {
+                                    if($category[0]->is_certificate_available($user_id)) {
```
Spacing

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1165,11 +1165,118 @@ public function returnCoursesAndSessions($user_id)
+                            }
+                        }
+                    }
+                    foreach($courses['not_category'] as $key => $courseNotInCatInfo) {
```
Spacing

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1165,11 +1165,118 @@ public function returnCoursesAndSessions($user_id)
+                            $courseNotInCatInfo['course_code'],
+                            array()
+                        );
+                        $courses['not_category'][ $key ]['student_info']['progress']  = ($progress === false)? null : $progress;
```
Two spaces before "=" sign -> should be only one space

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1165,11 +1165,118 @@ public function returnCoursesAndSessions($user_id)
+                        );
+                        $courses['not_category'][ $key ]['student_info']['certificate'] = null;
+                        if (isset($category[0])) {
+                            if($category[0]->is_certificate_available($user_id)) {
```
Spacing

ywarnier commented on main/inc/lib/userportal.lib.php:
```diff
@@ -1284,6 +1391,43 @@ public function returnCoursesAndSessions($user_id)
+
+                                        //Course option (show student progress)
+                                        //This code will add a new variables (Progress, Score, Certificate)
+                                        if(api_get_configuration_value('my_course_progress') === true && api_is_student() === true) {
```
Spacing

ywarnier commented on main/install/configuration.dist.php:
```diff
@@ -283,7 +283,9 @@
-// Set ConsideredWorkingTime work extra field variable from main/admin/extra_fields.php?type=work
+//Show student progress in My courses page
+//$_configuration['my_course_progress'] = false;
+// Set ConsideredWorkingTime work extra field variable to show in MyStudents page works report
```
Not sure if this is the result of a merge, but you shouldn't change the comments for another feature in your PR. This confuses things.

hi-rafa replied:
Yes this is the result of a merge.

ywarnier commented on main/install/data.sql:
```diff
@@ -1960,3 +1960,7 @@ INSERT INTO settings_current (variable, type, category, selected_value, title, c
+
+
+
+
```
Weird empty lines :-)

ywarnier commented on main/lang/catalan/trad4all.inc.php:
```diff
@@ -7986,4 +7986,9 @@
+$StudentCourseScore = "Puntuació: %s&#x00025;";
+$StudentCourseCertificate = "Certificat: %s";
+$MyCourseProgressTitle = "Mostra progrés del curs a l'estudiant";
+$MyCourseProgressTemplateComment = "En activar aquesta opció se'ls mostraran als estudiants el progrés dels seus cursos a la pàgina 'Els meus cursos'.";
```
Language files cannot be modified directly through a PR. You have to go through official developers or have admin status in translate.chamilo.org to add new variables.
When a language variable includes a %s sign, we name the variable with an uppercase "X" so that it is clear where the %s part is.
In this case, I have created the following variables for you (you will still need to connect to translate.chamilo.org to add the Catalan translations):
- $StudentCourseProgressX
- $StudentCourseScoreX
- $StudentCourseCertificateX
The two last variables should disappear for now as they won't be used until we move to the next major version and add these options to the settings_current table.
Please update your code to add the "X" letter at the end of your language variables.

hi-rafa commented:
Hi @ywarnier 
Thank you very much for you answer. The binary option would be something like this? Please let me know if you meant something else thanks.


/*
 * configuration.php
 */

//Show student progress in My courses page: 
//(100) = "certificate only"
//(010) = "score only"
//(001) = "progress only"

//Example
$_configuration['my_course_progress'] = "100";




/*
 * userportal.lib.php
 */

//(100) = "certificate only"
if (substr(api_get_configuration_value('my_course_progress'), 0,1)  ===  "1" ) {
       $specialCourses[$key]['student_info']['certificate'] = ........
}

//(010) = "score only"
if (substr(api_get_configuration_value('my_course_progress'), 1,1)  ===  "1" ) {
       $specialCourses[$key]['student_info']['score'] = ........
}

//(001) = "progress only"
if (substr(api_get_configuration_value('my_course_progress'), 2,1)  ===  "1" ) {
       $specialCourses[$key]['student_info']['progress'] = ........
}

ywarnier commented:
Sorry I didn't answer this very quickly. I actually meant an internal binary explanation, so if you just use integers from 0 to 7, we'll transform that into options after that.

hi-rafa commented:
Hi @ywarnier,
**Is there a specific reason why you would implement it that way?**
The code would look like this:

switch(api_get_configuration_value('my_course_progress')) {

case '000':
//Many lines .....
break;

case '100':
//Many lines .....
break;

case '001':
//Many lines .....
break;

case '010':
//Many lines .....
break;

case '110':
//Many lines .....
break;

..........

}

_By doing this aren't we repeating a lot of code?_

<hr>

**Why don't we instead add 3 variables?**
api_get_configuration_value('my_course_progress')
api_get_configuration_value('my_course_rate')
api_get_configuration_value('my_course_certificate')

ywarnier commented:
@hi-rafa, could you please change the contents of your branch to ensure all my comments are attended?

Regarding the numerical value: the idea is that all these options are part of the same setting. They all represent something that will appear at the same place and respond to the same visual style. In fact, they will all appear in the same "course-student-info" box (so the general setting should probably be called 'course_student_info' instead of course_progress, which is one of the sub-options, but it's true that we could do that differently.

Because of that reason above, we will try to show them all inside *one* setting, like it's the case for the whoisonline setting:
![whoisonline-setting](https://cloud.githubusercontent.com/assets/609839/24410187/0e021a1e-1398-11e7-9e7e-f90f48966b0b.png)

In the database, we have to store that as *one* record in the settings_current table, and multiple options in the settings_options table. As such, the value for the setting will be recovered through a call like this (only when it is integrated to the database in version 2.0):
```
$showStudentScore = api_get_setting('course_student_info', 'score');
$showStudentProgress = api_get_setting('course_student_info', 'progress');
$showStudentCertificate = api_get_setting('course_student_info', 'certificate');
```
So I was thinking if we represented it as a binary number it would be easier for now, but actually there might be a better way.
We could simply store it like this (in configuration.php):
```
$_configuration['course_student_info']['score'] = 'true';
$_configuration['course_student_info']['progress'] = 'true';
$_configuration['course_student_info']['certificate'] = 'true';
```
And then get it from anywhere with:
```
$showStudentScore = api_get_configuration_value('course_student_info')['score'];
```
(maybe with a few additional checks to avoid an E_NOTICE.

I think that would be better. Thanks for doubting me, I was thinking only about the migration process.

ywarnier commented:
Your PR has been included in Chamilo. It will be part of 1.11.4.
I have adapted it a bit to reduce calls to api_get_configuration_value().
You will see that, as a result of the cleanup of the translation files, your StudentCourseCertificateX variable is not present in Catalan anymore. As such, it will probably take the default value of the Spanish translation. Just go to translate.chamilo.org to update this term (let me know if your account doesn't get approved) and it will be part of the language pack of 1.11.4.

ywarnier commented:
Thank you for your contribution :+1:  (I forgot that part)

hi-rafa commented:
Hi @ywarnier 
My translate.chamilo.org account has not been approved. My username is teran.gautier.

Thank you

ywarnier commented:
I approved one "rafa-something" overnight (only for translations from english to catalan, though, as others are covered). Please check and let me know.

hi-rafa commented:
My username is teran.gautier. By the way the Password Recovery system of translate.chamilo is not working.

ywarnier commented:
Enabled. Set your password = your username for now. Please change it :-)
We're abandoning translate.chamilo.org after version 1.11 (in favour of crowd.in), so not really important for us to fix the password reminder feature
